### DISCUSSION TITLE
Dependency Injection and modifying the generated code

### DISCUSSION THREAD

**Discussion by Southclaws** [prisma/prisma#5895] [Q&A] [answered]
Hi! I've been using Prisma with Go for a while and love it. I just started a new toy project in TypeScript to get to know the language a bit more on the backend and was playing with some DI frameworks. I tried to use tsyringe with Prisma but I think I ran into issues because the generated class needs `@injectable()` decorating it.

The problem is, Prisma's code is generated so I shouldn't be editing it as it's not sustainable long-term to keep doing that.

But I also know that Prisma's docs recommend simply instantiating a single instance in a module and exporting it for use globally.

What's the best practice here? Would Prisma benefit from being able to play well with DI frameworks or is it just not worth the hassle?

Thanks!

**Comment by ryands17** [comment]
@Southclaws 👋 

The best way is to use a single `PrismaClient` instance and import that wherever necessary. I am not familiar with `tsyringe` but even if you're using DI, I would suggest creating a your own wrapper class, instantiating `PrismaClient` in it and just using it as a singleton class. That would be the preferred approach.

**Comment by Southclaws** [comment, marked as answer]
I solved this by using a named token rather than deriving it from the PrismaClient type, and using `useValue` to manually construct a single instance for use throughout the app. For some reason, tsyringe could not figure out how to construct the client but this works fine:

```ts
container.register<PrismaClient>("PrismaClient", {
  useValue: new PrismaClient(),
});
```

So now constructors of dependent types can use:

```ts
  constructor(
    @inject("PrismaClient") private db: PrismaClient
  ) { ... }
```

**Reply by Farki** [reply]
Doesn't work for me: 
```
Cannot inject the dependency "prismaClient" at position #0 of "CustomerRepository" constructor. Reason:
TypeInfo not known for "PrismaClient"
```

**Reply by Southclaws** [reply]
Make sure container.register is called before any instances of the class are instantiated

**Reply by Farki** [reply]
Thanks for reply, but my mistake was in token. I used class PrismaClient instead of string "PrismaClient".

**Reply by HighLiuk** [reply]
`Decorators are not valid here. ts(1206)`

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
